docs/guides/user-containers/readme.md

```diff
@@ -262,16 +262,106 @@ allowed per workflow.
-level workings.
+- To ensure the proper permissions (i.e. user/group) are configured
+- For MPI jobs, to ensure the launcher pod can contact the worker pods
+
+### PostRun
+
+In PostRun, the containers are expected to exit cleanly with a zero exit code. If a container fails to
+exit cleanly, the Kubernetes software attempts a number of retries based on the
+configuration of the container profile. It continues to do this until the container exits
+successfully, or until the `retryLimit` is hit - whichever occurs first. In the latter case, the
+workflow reports an Error.
+
+Read up on the [Failure Retries](#failure-retries) for more information on retries.
+
+Furthermore, the container profile features a `postRunTimeoutSeconds` field. If this timeout is
+reached before the container successfully exits, it triggers an Error state. The timer for this
+timeout begins upon entry into the PostRun phase, allowing the containers the specified period to
+execute before the workflow enters an Error state.
+
+To recap the PostRun behavior:
+
+- If the container exits successfully, transition to Ready:true.
+- If the container exits unsuccessfully after `retryLimit` number of retries, transition to the
+Error State.
+- If the container is running and has not exited after `postRunTimeoutSeconds` seconds, terminate
+the container and transition to the Error State.
+
+### Failure Retries
+
+If a container fails (non-zero exit code), the Kubernetes software implements retries. The number of
+retries can be set via the `retryLimit` field in the container profile. If a non-zero exit code is
+detected, the Kubernetes software creates a new instance of the pod and retries. The
+default number of retries for `retryLimit` is set to 6, which is the default value for Kubernetes
+Jobs. This means that if the pods fails every single time, there will be 7 failed pods in total
+since it attempted 6 retries after the first failure.
+
+To understand this behavior more, see [Pod backoff failure
+policy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-backoff-failure-policy)
+in the Kubernetes documentation. This explains the retry (i.e. backoff) behavior in more detail.
+
+It is important to note that due to the configuration of the `MPIJob` and/or `Job` that is created
+for User Containers, the container retries are immediate - there is no backoff timeout between
+retires. This is due to the NNF Software setting the `RestartPolicy` to `Never`, which causes a new
+pod to spin up after every failure rather than re-use (i.e. restart) the previously failed pod. This
+allows a user to see a complete history of the failed pod(s) and the logs can easily be obtained.
+See more on this at [Handling Pod and container
+failures](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
+in the Kubernetes documentation.
 
 ## Putting it All Together
```

Comment:
I'm missing a piece: How does the user container know when PostRun begins?  Where does it get that cue?

Reply:
Is this regarding the PostRun timeout? The container has no idea - the NNF software sets the Job's activeDeadline when the workflow transitions to PostRun. Once that activeDeadline timeout is hit, k8s takes care of the rest and the containers are killed. The job transitions to a failed state and the NNF software turns that into a workflow error.

Reply:
Is the user container supposed to do some work during PostRun?

Reply:
That's up to the application developer that created the image, but they won't have any cue from the NNF software that it's in PostRun.

I envision the port being the mechanism that drives behavior to the application (e.g. SCR) running in the container. The compute node application contacts the application in the container and tells it that it's time to do something.

1. In PreRun, user container starts successfully. The workflow transitions to PreRun complete
2. Flux runs the application on the compute node
3. The compute node application contacts the user container via the port to drive behavior, ultimately saying: "I'm done"
4. Compute node application finishes and Flux sees that
5. Flux transitions to PostRun
6. PostRun expects the user container to exit cleanly before transitioning to PostRun complete

--- 

@roehrich-hpe and I discussed this at my desk already, but posting this for posterity.

I'll also be adding a section that illustrates this more directly.